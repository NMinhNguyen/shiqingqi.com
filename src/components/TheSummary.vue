<template>
  <div class="the-summary" :style="summaryStyle">
    <p class="the-summary__i-believe">
      {{ $store.getters.getText("SUMMARY_LINE_1") }}
    </p>
    <p class="the-summary__craftsman">
      {{ $store.getters.getText("SUMMARY_LINE_2") }}
    </p>
  </div>
</template>

<script>
export default {
  props: {
    animateLineOne: { type: Number, default: 0 },
    animateLineTwo: { type: Number, default: 0 }
  },
  computed: {
    /** @type {() => {[property: string]: number}} */
    summaryStyle() {
      return {
        "--the-summary-animate-line-1": this.animateLineOne,
        "--the-summary-animate-line-2": this.animateLineTwo
      };
    }
  }
};
</script>

<style lang="scss">
.the-summary {
  text-align: center;
  padding: 30vh 3rem;

  &__i-believe {
    $animate: var(--the-summary-animate-line-1);
    transform: translateY(calc((1 - #{$animate}) * 20vh));
    opacity: $animate;
    font-family: $font-heading;
    font-size: 2rem;
  }

  &__craftsman {
    $animate: var(--the-summary-animate-line-2);
    transform: translateY(calc((1 - #{$animate}) * 15vh));
    opacity: $animate;
    font-size: 1.3rem;
  }
}
</style>
