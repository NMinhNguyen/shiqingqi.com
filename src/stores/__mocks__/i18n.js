export default {
  state: {
    locale: "en",
    translations: {
      en: {
        test: "test"
      }
    }
  },
  mutations: {
    setLocale: jest.fn(),
    setTranslations: jest.fn()
  },
  actions: {
    fetchTranslations: jest.fn()
  },
  getters: {
    getText: jest.fn(() => () => "")
  }
};
