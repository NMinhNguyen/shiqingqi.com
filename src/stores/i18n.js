import Vue from "vue";

export const state = {
  locale: "en",
  /** @type {{[locale:string]: {[key:string]: string}}} */
  translations: {}
};

export const mutations = {
  /**
   * @param {typeof state} state State object
   * @param {string} locale Locale code
   */
  setLocale(state, locale) {
    state.locale = locale;
  },

  /**
   * @param {typeof state} state State object.
   * @param {{lang: string, translations: object}} payload New translations.
   */
  setTranslations(state, { lang, translations }) {
    Vue.set(state.translations, lang, {
      ...state.translations[lang],
      ...translations
    });
  }
};

export const actions = {
  /**
   * @param {import("vuex").ActionContext<typeof state, {}>} context
   * @param {string} lang Language code.
   * @returns {Promise}
   */
  async fetchTranslations({ commit }, lang) {
    /** @type {{default: object}} */
    const module = await import(`@/constants/translations/${lang}.json`);
    commit("setTranslations", { lang, translations: module.default });
  }
};

export const getters = {
  /**
   * @param {typeof state} state
   * @returns {(key: string, params: {[key:string]: string}) => string}
   */
  getText: state => (key, params) => {
    const translations = state.translations[state.locale];

    // Return empty string if translation doesn't exist
    if (!translations || !translations[key]) return "";

    // Return translation if no params provided
    if (!params) return translations[key];

    // Replace the params
    let text = translations[key];
    for (let name in params) {
      text = text.replace(`{{${name}}}`, params[name]);
    }
    return text;
  }
};

export default {
  state,
  mutations,
  actions,
  getters
};
