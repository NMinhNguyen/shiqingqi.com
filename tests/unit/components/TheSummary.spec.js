// @ts-nocheck
import { shallowMount, createLocalVue } from "@vue/test-utils";
import Vuex from "vuex";
import TheSummary from "@/components/TheSummary.vue";
import i18n from "@/stores/i18n";

jest.mock("@/stores/i18n");

let localVue = createLocalVue();
localVue.use(Vuex);

describe("TheSummary", () => {
  let store;

  beforeAll(() => {
    store = new Vuex.Store({
      modules: {
        i18n
      }
    });
  });

  it("renders paragraphs", () => {
    expect(
      shallowMount(TheSummary, {
        localVue,
        store
      })
    ).toMatchSnapshot();
  });
});
