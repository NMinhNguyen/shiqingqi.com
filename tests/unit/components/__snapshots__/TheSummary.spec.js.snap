// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TheSummary renders paragraphs 1`] = `
<div class="the-summary">
  <p class="the-summary__i-believe">

  </p>
  <p class="the-summary__craftsman">

  </p>
</div>
`;
